### PR TITLE
[BEAM-923] Add localhost option for DatastoreIO (fix assignment issue)

### PR DESCRIPTION
**Pull request by iemejia**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by iemejia** [comment]
R: @vikkyrk
and @davorbonaci for the merge.

**Comment by vikkyrk** [comment]
@iemejia Thanks for catching the issue. 
One thing to note here is that this will be a breaking user facing change. The reason we have `Write`, `DeleteEntity` and `DeleteKey` classes is to be more explicit about what the mutation is doing. So its better to keep as is but still submit the `localhost` fix.

**Comment by iemejia** [comment]
I agree that this is a breaking change for the return types, but don't you think this is better? Independently of the different types the whole hierarchy didn't make sense, and also remember that the DataStoreIO API is marked as experimental and can still evolve, and this is probably the best moment to make these changes before the stable API no ?

**Comment by iemejia** [comment]
Notice also that I am not breaking the global signature, users still access the PTransform with the same methods:
DatastoreIO.v1().write(), DatastoreIO.v1().deleteEntity() and DatastoreIO.v1().deleteKey()

**Comment by vikkyrk** [comment]
@iemejia The `Experimental` had to be removed long time ago but is being removed here https://github.com/apache/beam/pull/1369. 

Agreed that it will not change the signature for regular usages, it will still break for something like:
`DatastoreV1.DeleteEntity delete = p.apply(...);`

But I remember the main reason we did not go with `Mutate` approach is because it allows mix and match of delete and write mutations which can suboptimal wrt. Cloud Datastore when done in bulk.

**Comment by vikkyrk** [comment]
(Btw I agree in general that `Mutate` is better approach abstraction wise)

**Comment by iemejia** [comment]
Can I move the two methods into the parent at least (to avoid the repeated code).
withProjectId(String projectId) and  withLocalhost(String localhost) ?

**Comment by vikkyrk** [comment]
I am fine although the problem with that is `Mutate` is private, so not sure if it will work. But if you can get it to work that would be nice!

**Comment by iemejia** [comment]
Yes, it was not possible to do this, actually the problem is the return type of those types. I just fixed respecting the types, hope it is ok now.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java [hunk 1]
```diff
@@ -796,9 +796,9 @@ public DeleteKey withLocalhost(String localhost) {
    */
   private abstract static class Mutate<T> extends PTransform<PCollection<T>, PDone> {
     @Nullable
-    private final String projectId;
+    final String projectId;
```
Comment (vikkyrk) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: Keep the fields `private` and use `gettter` methods instead in the sub-classes?

Comment (iemejia) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: Yes you are right, seems a bit convoluted, but ok, fixing again.

**Comment by iemejia** [comment]
Fixed and rebased, hope it is ok now.

**Comment by vikkyrk** [comment]
Given that `localhost` option doesn't have integration tests, this LGTM.

Need not be in this PR, but in light of the recent discussion on `Style: how much testing for transform builder classes?` few things we could also do:

- Convert `Write`/`Delete` to use `AutoValue`
- Remove null checks and tests for `withOption` methods. 

cc @jkff

**Comment by jkff** [comment]
Oh I see, this is the PR you mentioned on the style thread.

Indeed the proper way to fix this is to have integration tests. If that is not feasible for some reason (or outside the scope of this PR), then yes - convert these classes to AutoValue so that such typo errors become extremely unlikely (another benefit of AutoValue).

**Pull request closed without merging**